eclair-core/src/main/scala/fr/acinq/eclair/Features.scala

```diff
@@ -19,63 +19,123 @@ package fr.acinq.eclair
+  case object OptionDataLossProtect extends Feature {
+    val rfcName = "option_data_loss_protect"
+    val mandatory = 0
+    val optional = 1
+  }
 
-  val CHANNEL_RANGE_QUERIES_BIT_MANDATORY = 6
-  val CHANNEL_RANGE_QUERIES_BIT_OPTIONAL = 7
+  case object InitialRoutingSync extends Feature {
+    val rfcName = "initial_routing_sync"
+    // reserved but not used as per lightningnetwork/lightning-rfc/pull/178
+    val mandatory = 2
+    val optional = 3
+  }
 
-  val VARIABLE_LENGTH_ONION_MANDATORY = 8
-  val VARIABLE_LENGTH_ONION_OPTIONAL = 9
+  case object ChannelRangeQueries extends Feature {
+    val rfcName = "gossip_queries"
+    val mandatory = 6
+    val optional = 7
+  }
 
-  val CHANNEL_RANGE_QUERIES_EX_BIT_MANDATORY = 10
-  val CHANNEL_RANGE_QUERIES_EX_BIT_OPTIONAL = 11
+  case object VariableLengthOnion extends Feature {
+    val rfcName = "var_onion_optin"
+    val mandatory = 8
+    val optional = 9
+  }
 
-  val PAYMENT_SECRET_MANDATORY = 14
-  val PAYMENT_SECRET_OPTIONAL = 15
+  case object ChannelRangeQueriesExtended extends Feature {
+    val rfcName = "gossip_queries_ex"
+    val mandatory = 10
+    val optional = 11
+  }
 
-  val BASIC_MULTI_PART_PAYMENT_MANDATORY = 16
-  val BASIC_MULTI_PART_PAYMENT_OPTIONAL = 17
+  case object PaymentSecret extends Feature {
+    val rfcName = "payment_secret"
+    val mandatory = 14
+    val optional = 15
+  }
+
+  case object BasicMultiPartPayment extends Feature {
+    val rfcName = "basic_mpp"
+    val mandatory = 16
+    val optional = 17
+  }
 
   // TODO: @t-bast: update feature bits once spec-ed (currently reserved here: https://github.com/lightningnetwork/lightning-rfc/issues/605)
-  // We're not advertizing these bits yet in our announcements, clients have to assume support.
+  // We're not advertising these bits yet in our announcements, clients have to assume support.
   // This is why we haven't added them yet to `areSupported`.
-  val TRAMPOLINE_PAYMENT_MANDATORY = 50
-  val TRAMPOLINE_PAYMENT_OPTIONAL = 51
+  case object TrampolinePayment extends Feature {
+    val rfcName = "trampoline_payment"
+    val mandatory = 50
+    val optional = 51
+  }
+
+  // Features may depend on other features, as specified in Bolt 9.
+  private val featuresDependency = Map(
+    ChannelRangeQueriesExtended -> (ChannelRangeQueries :: Nil),
+    PaymentSecret -> (VariableLengthOnion :: Nil),
+    BasicMultiPartPayment -> (PaymentSecret :: Nil),
+    TrampolinePayment -> (PaymentSecret :: Nil)
+  )
+
+  case class FeatureException(message: String) extends IllegalArgumentException(message)
+
+  def validateFeatureGraph(features: BitVector): Option[FeatureException] = featuresDependency.collectFirst {
+    case (feature, dependencies) if hasFeature(features, feature, None) && dependencies.exists(d => !hasFeature(features, d, None)) =>
+      FeatureException(s"${features.toBin} sets $feature but is missing a dependency (${dependencies.mkString(" or ")})")
```

Comment:
A bit more expensive, but much more friendly:

```suggestion
      FeatureException(s"${features.toBin} sets $feature but is missing a dependency (${dependencies.filter(d => !hasFeature(features, d, None)).mkString(" and ")})")
```